src/commissioner/Discoverer.cpp

```diff
@@ -0,0 +1,264 @@
+ *    limitations under the License.
+ */
+
+#include <commissioner/Discoverer.h>
+
+namespace chip {
+namespace Commissioner {
+namespace CommissionableNodeDiscoverer {
+
+#if CONFIG_NETWORK_LAYER_BLE
+BleDiscoverer::BleDiscoverer(Controller::DeviceControllerSystemState & systemState, Delegate ** delegate) :
+    Joinable(delegate), mSystemState(systemState)
+{
+    mSystemState.Retain();
+}
+
+BleDiscoverer::~BleDiscoverer()
+{
+    mSystemState.Release();
+}
+
+void BleDiscoverer::Shutdown()
+{
+    if (InProgress())
+    {
+        VerifyOrReturn(mSystemState.BleLayer()->CancelBleIncompleteConnection() == CHIP_NO_ERROR);
+        ReleaseShutdownToken();
+    }
+}
+
+CHIP_ERROR BleDiscoverer::StartBleDiscovery(SetupPayload & payload, Platform::SharedPtr<ShutdownToken> token)
+{
+    VerifyOrReturnError(token.get() != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    ReturnErrorOnFailure(mSystemState.BleLayer()->NewBleConnectionByDiscriminator(payload.discriminator, this,
+                                                                                  OnBleDiscoverySuccess, OnBleDiscoveryError));
+    RetainShutdownToken(token);
+    return CHIP_NO_ERROR;
+}
+
+void BleDiscoverer::TallyBleDiscovery(BLE_CONNECTION_OBJECT connection)
+{
+    mBleConnection.SetValue(connection);
+    (*mDelegate)->OnDiscovery();
+    ReleaseShutdownToken();
+}
+
+void BleDiscoverer::TallyBleFailure()
+{
+    (*mDelegate)->OnDiscovery();
+    ReleaseShutdownToken();
+}
+
+CHIP_ERROR BleDiscoverer::GetNextBleCandidate(Commissionee & commissionee)
+{
+    VerifyOrReturnError(mBleConnection.HasValue(), CHIP_ERROR_NOT_FOUND);
+    commissionee.CloseBle(); // close if connection currently exists; failure is a no-op
+    commissionee.mBleConnection = mBleConnection;
+    commissionee.mCommissionableNodeAddress.SetValue(Transport::PeerAddress::BLE());
+    commissionee.mMrpConfig.ClearValue();
+    ReturnErrorOnFailure(commissionee.OpenBle());
+    mBleConnection.ClearValue();
+    ChipLogProgress(Controller, "Commissionable node candidate identified via BLE");
+    return CHIP_NO_ERROR;
+}
+
+void BleDiscoverer::OnBleDiscoverySuccess(void * context, BLE_CONNECTION_OBJECT connection)
+{
+    BleDiscoverer * instance = static_cast<BleDiscoverer *>(context);
+    instance->TallyBleDiscovery(connection);
+}
+
+void BleDiscoverer::OnBleDiscoveryError(void * context, CHIP_ERROR err)
+{
+    ChipLogError(Controller, "BLE Commissionable Node Discovery failed: %s", ErrorStr(err));
+    BleDiscoverer * instance = static_cast<BleDiscoverer *>(context);
+    instance->TallyBleFailure();
+}
+
+#endif // CONFIG_NETWORK_LAYER_BLE
+
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+DnssdDiscoverer::DnssdDiscoverer(Controller::DeviceControllerSystemState & systemState, Delegate ** delegate) :
+    Joinable(delegate), mSystemState(systemState)
+{
+    mSystemState.Retain();
+}
+
+DnssdDiscoverer::~DnssdDiscoverer()
+{
+    mSystemState.Release();
```

Comment:
Would it make sense to call `::Shutdown()` here as well?

Reply:
It wouldn't, and this shows that I need to add documentation to src/commissioner/Discoverer.h.

The intended operation of Discoverer is:

* exist as shared_ptr member variable of a VariantState object
* provide async Shutdown method that can 'join' all internal resources back to Discoverer (Joinable interface)
* when all resources have joined, call OnShutdown delegate interface
* release other, dependency-injected resources on destruct
* destruct happens automatically when the last state releases its shared_ptr reference

This is a good approach because different discovery interfaces have different cancellation capabilities on different platforms.  For instance, BLE discovery cannot currently be canceled at all on Linux platform.  And all of this will be further complicated when SoftAP discovery is added post-1.0.

The Joinable interface + shared_ptr solves a pretty difficult problem: the state machine needs to be able to tell a Discoverer doing multiple concurrent things when the state machine is done Discovering (with Shutdown method), but the Discoverer needs to be able to perform Shutdown asynchronously over some indeterminate period of time.  As the Discoverer is doing this, it will be cleaning up multiple, concurrent resources.

I will keep this comment open until I've added sufficient documentation to src/commissioner/Discoverer.h.

Reply:
I've added comment headers explaining this operation.